lighthouse-core/gather/gatherers/seo/tap-targets.js

```diff
@@ -94,6 +99,7 @@ function getVisibleClientRects(element) {
   // - tap targets audit doesn't consider different containers/layers
   // - having most content in an explicit scroll container is rare
   // - treating them as hidden only generates false passes, which is better than false failures
+  // TODO: is this still needed?
```

Comment:
Right now only need this to filter out rects that are outside the visible scroll area, and which end up being the largest rect. Without `filterClientRectsWithinAncestorsVisibleScrollArea` we'll fail to detect those tap targets.

Maybe that's fine? If we didn't filter that rect out, and that rect did overlap with another tap target, then the failure would be invalid.

Maybe we shouldn't check the largest rect of each tap target with `elementFromPoint`, but rather filter the list of client rects calling `elementFromPoint` on each one?

Reply:
> Maybe we shouldn't check the largest rect of each tap target with elementFromPoint, but rather filter the list of client rects calling elementFromPoint on each one?

I'm not sure I understand the proposal. What's the distinction between that and what we do now? what would start failing/passing?

Reply:
Current approach in this PR:

- get all client rects that aren't 0x0px or outside the scroll ancestor
- get the largest of those rects
- the element is visible if `getElementFromPoint` at the center of the largest rect matches the element (but the client rect list also contains rects where the center is invisible)

Proposal:

- get all client rects that aren't 0x0px (or outside the scroll ancestor)
- each rect is visible if `getElementFromPoint` at its center matches the element
- the element is visible if it has a visible rect

With the current approach we fail to detect the tap targets in the [GoDaddy example](https://gist.github.com/mattzeunert/7e1d8f3bfd2ce7aa29ab1b4e6790966c). That's because the center of the largest rect is invisible, so we decide that the whole element must be invisible. By moving to the new approach we fix the GoDaddy problem.

I was also thinking we might be able to get rid of `filterClientRectsWithinAncestorsVisibleScrollArea`, but if we just check center of each react we still risk false failures caused by partially invisible rects inside `overflow: hidden` containers. So maybe we should test all four corners with `getElementFromPoint`?

Let me know if you have any thoughts. I think I need to do some more experimentation, and also see what the performance cost of `elementFromPoint` and `scrollTo` is.

Reply:
I've changed it to do the z index test on each client rect now.

I don't think we need `filterClientRectsWithinAncestorsVisibleScrollArea` any more, because the z index check covers most of those cases. The only difference is that z index only checks the center of the tap target whereas before we were checking that the whole rect is visible.

Stats for this [wiki page](https://en.m.wikipedia.org/wiki/List_of_2017_albums):

|    | master | this branch |
| - | ------- | ----------- |
| Gathered tap targets | 2018 |  2091 |
| Run time | 405ms | 497ms |